bbot/modules/deadly/medusa.py

```diff
@@ -0,0 +1,234 @@
+                "cmd": "make",
+            },
+        },
+        {
+            "name": "Build medusa: make install",
+            "command": {
+                "chdir": "#{BBOT_TEMP}/medusa/workdir",
+                "cmd": "make install",
+                "creates": "#{BBOT_TEMP}/medusa/build/bin/medusa",
+            },
+        },
+        {
+            "name": "Install medusa",
+            "copy": {
+                "src": "#{BBOT_TEMP}/medusa/build/bin/medusa",
+                "dest": "#{BBOT_TOOLS}/",
+                "mode": "u+x,g+x,o+x",
+            },
+        },
+    ]
+
+    async def setup(self):
+        # Try to cache wordlist
+        try:
+            self.snmp_wordlist_path = await self.helpers.wordlist(self.config.get("snmp_wordlist"))
+        except WordlistError as e:
+            return False, f"Error retrieving wordlist: {e}"
+
+        return True
+
+    async def filter_event(self, event):
+        handled_protocols = ["snmp"]  # Could be extended later
+
+        protocol = event.data["protocol"].lower()
+        if not protocol in handled_protocols:
+            return False, f"service {protocol} is currently not supported. Only SNMP."
+
+        return True
+
+    async def handle_event(self, event):
+        host = str(event.host)
+        port = str(event.port)
+        protocol = event.data["protocol"].lower()
+
+        if protocol == "snmp":
+            snmp_versions = self.config.get("snmp_versions")
+
+            # Medusa must be called for each SNMP version separately after each run finished.
+            for snmp_version in snmp_versions:
+                command = await self.construct_command(host, port, protocol, snmp_version)
+
+                result = await self.run_process(command)
+
+                if result.stderr:
+                    # Medusa outputs to stderr if a readonly community was found in WRITE mode
+                    # That's intended behavior
+                    self.info(f"Medusa stderr: {result.stderr}")
+
+                async for message in self.parse_output(result.stdout, event, snmp_version):
+                    vuln_event = self.create_vuln_event("CRITICAL", message, event)
+                    await self.emit_event(vuln_event)
+
+        # else: Medusa supports various protocols which could in theory be implemented later on.
+
+    async def parse_output(self, output, event, protocol_version):
+        for i, line in enumerate(output.splitlines()):
+            # Print original Medusa output
+            self.info(line)
+
+            if "FOUND" in line:
+                # Some credential was guessed
+                password_match = re.search(r"Password:\s*(\S+)", line)
+                password = password_match.group(1) if password_match else None
+
+                success_indicator_match = re.search(
+                    r"\[([^\]]+)\]\s*$", line
+                )  # Captures text inside last brackets at end
```

Comment:
One more thing when you go to lint; can we compile these regexes in .setup() to save on CPU? Thanks.

Reply:
Done!